apps/discussions/contracts/DiscussionApp.sol

```diff
@@ -0,0 +1,125 @@
+
+    struct DiscussionPost {
+        address author;
+        string postCid;
+        uint discussionThreadId;
+        uint id;
+        uint createdAt;
+        bool show;
+        string[] revisionCids;
+    }
+
+    uint discussionThreadId;
+
+    mapping(uint => DiscussionPost[]) public discussionThreadPosts;
+
+    function initialize() public onlyInit {
+        discussionThreadId = 0;
+        initialized();
+    }
+
+    /**
+     * @notice Create discussion post with an IPFS content hash '`postCid`'.
+     * @param postCid The IPFS content hash of the discussion post data
+     * @param discussionThreadId The thread to post this discussion to
+     */
+    function post(string postCid, uint discussionThreadId) external auth(DISCUSSION_POSTER_ROLE) {
+        DiscussionPost storage post;
+        post.author = msg.sender;
+        post.postCid = postCid;
+        post.discussionThreadId = discussionThreadId;
+        post.createdAt = now; // solium-disable-line security/no-block-members
+        post.show = true;
+        uint postId = discussionThreadPosts[discussionThreadId].length;
+        post.id = postId;
+        discussionThreadPosts[discussionThreadId].push(post);
+        emit Post(msg.sender, postCid, discussionThreadId, postId, now); // solium-disable-line security/no-block-members
+    }
+
+    /**
+     * @notice Hide a discussion post with ID '`postId`'.
+     * @param postId The postId to hide
+     * @param discussionThreadId The thread to hide this discussion from
+     */
+    function hide(uint postId, uint discussionThreadId) external auth(DISCUSSION_POSTER_ROLE) {
+        DiscussionPost storage post = discussionThreadPosts[discussionThreadId][postId];
+        require(post.author == msg.sender, "You cannot hide a post you did not author.");
+        post.show = false;
+        emit Hide(msg.sender, discussionThreadId, postId, now); // solium-disable-line security/no-block-members
+    }
+
+    /**
+     * @notice Revise a discussion post with ID '`postId`'.
+     * @param revisedPostCid The cid of the pre-revised post
+     * @param postId The postId to revise
+     * @param discussionThreadId The thread to hide this discussion from
+     */
+    function revise(string revisedPostCid, uint postId, uint discussionThreadId) external auth(DISCUSSION_POSTER_ROLE) {
+        DiscussionPost storage post = discussionThreadPosts[discussionThreadId][postId];
+        require(post.author == msg.sender, "You cannot revise a post you did not author.");
+        // add the current post to the revision history
+        // should we limit the number of revisions you can make to save storage?
+        post.revisionCids.push(post.postCid);
+        post.postCid = revisedPostCid;
+        emit Revise(
+            msg.sender,
+            revisedPostCid,
+            discussionThreadId,
+            postId,
+            post.createdAt,
+            now // solium-disable-line security/no-block-members
```

Comment:
This is redundant since the time information should be accessible from the event either way.